アクセストークン用のモデルを追加

a5ob7r opened the pull request:
# 概要

OAuth 2.0のアクセストークンを表現するモデルを追加する。

- アクセストークンはユーザに対して複数発行されうる
- トークン文字列やスコープ、有効期限、失効されているかどうかなどの情報が設定される
- 単純なスコープの設定なので単なるスペース区切りの文字列として直接保存する。
    - スペース区切りの文字列という表現自体はOAuth 2.0でのスコープの表現と同様である。
    - https://datatracker.ietf.org/doc/html/rfc6749#section-3.3
    - 要件に応じてスコープはどこまでも複雑にできそうだが、今回はトレーニング目的なのでそこまでやる必要はないと判断した。

a5ob7r committed:
Add AccessToken model

tk0miya commented on spec/models/access_token_spec.rb:
```diff
@@ -0,0 +1,176 @@
+require 'rails_helper'
+
+RSpec.describe AccessToken, type: :model do
+  subject { described_class.new token:, scope:, expires_in:, revoked_at:, user: }
```
AccessToken.new するときにトークン文字列を外部から指定するんですか?
また、new するときに revoked_at を指定することはありえますか?

(他の scope, expires_in, user はなんとなく予想がついた)

a5ob7r replied:
そういえば、他のモデルやコントローラがどうやってAccessTokenインスタンスを作るか考えてなかったです。

クラスメソッドあたりを書くのかな、などと考え中です。

tk0miya replied:
開発しているクラスやメソッドがどの用に使われるのか、というのが spec で表現できるといいですよね。

普段はつい「spec == テスト」と考えがちですが、spec はあくまで「動く仕様(spec)」であって、主は仕様です。
モデルのようなクラスの場合は「機能 == どう使えるのか」がポイントになるのです。

Taku3939 commented on spec/factories/access_tokens.rb:
```diff
@@ -0,0 +1,10 @@
+  factory :access_token do
+    token { SecureRandom.base64 }
+    scope { 'READ WRITE' }
+    expires_in { 60 * 60 }
```
1.hourの方が見やすいかも

a5ob7r replied:
おー、Railsっぽい。ありがとうございます。変更します。

a5ob7r committed:
Add a format validation for a token string

This format contains string length validation too.

a5ob7r committed:
Change a test data for a token string

a5ob7r committed:
Switch to ActiveSupport::Duration to specify relative time

tim-toga commented on app/models/access_token.rb:
```diff
@@ -0,0 +1,26 @@
+  end
+
+  def revoked?
+    !revoked_at.nil?
```
revoked_at.present? の方が分かりやすいですね

https://railsdoc.com/page/empty
present? == !blank? だから単純な置き換えという訳ではないけど、この場面ではpresent?でいいはず

a5ob7r replied:
なるほど、修正します。

tim-toga commented on spec/models/access_token_spec.rb:
```diff
@@ -0,0 +1,189 @@
+      end
+
+      it 'trueを返す' do
+        travel_to Date.current.tomorrow do
```
travel_to はじめてみた。こうすると`Time.current`が明日になるのか。便利だ

epaew replied:
「`travel_to` を使う」こと自体はいいのだけれど、
- `'失効日時が過去のとき'` って context との相性
- `it` ブロック内で `travel_to` していること

は気になりますねえ。

---
（おまけ）

そもそも「失効日時」は「`created_at + expires_in`」のことを指しているみたいだけど、これって「`#expires?` 内にのみ存在する＝ただの内部表現」ですよね。  
その「内部表現」を閾値にしてのテストって、どうなんだろう？

tk0miya replied:
@tim-toga 以前は timecop を使うのが一般的だったけど、いくつか前の Rails から ActiveSupport に travel_to が収録されるようになったので、わざわざ timecop gem を入れる必要はなくなりましたね。

tk0miya replied:
「失効日時が過去のとき」という context に一致するよう、travel_to は before 節に書くのがよいですね。

travel_to は before 節で呼び出した場合は example の最後に自動的に終了処理をしてくれるので
```
before { travel_to Date.current.tomorrow }
```

で良いはずです。
https://shinkufencer.hateblo.jp/entry/2019/09/22/233000

tk0miya replied:
…とコメントしてから気づいたのですが、「失効日時が過去のとき」であれば、過去の失効日時(revoked_at)を指定するのが context と一致しているのでは?
発行してから時間が未来に進んだとき、というのは確かに「失効日時が過去のとき」ではあるのですが、間接的な表現で分かりづらいですよね。

tk0miya replied:
あ、revoked_at じゃないか。created_at + expires_in が起点なので、過去のトークンを作るって手もありそうです、と言いたかった。

a5ob7r replied:
> - '失効日時が過去のとき' って context との相性
> - it ブロック内で travel_to していること
>
> は気になりますねえ。

ご指摘のとおりでしたので、#travel_toをbeforeブロック内に移動させて、かつcontextの説明に合致させるため、すでに失効しているアクセストークンを作るように修正しました。おかげでコードもすっきりしました。

a5ob7r committed:
Add a missing association to User about AccessToken

a5ob7r committed:
Use `#present?` instead of not `#nil?`

epaew commented on app/models/access_token.rb:
```diff
@@ -0,0 +1,26 @@
+  validates :scope, format: { with: /\A(READ|WRITE)( (READ|WRITE)*)\Z/ }
+  validates :expires_in, numericality: { greater_than_or_equal_to: 0 }
+
+  def expires?
```
[IMO]
1. このメソッドの名前は `#expired?` のほうがいい気がする

a5ob7r replied:
.expires_inに引っ張られていましたが、確かにそのほうが自然ですね。これだと、いついつまでに期限が切れるかを調べることになるのかな。修正します。

a5ob7r replied:
やるとしたらこんな感じですかね。

```ruby
  def expires_at?(datetime)
    created_at + expires_in < datetime
  end
```

tk0miya replied:
このメソッドがどのように使われるのか、考えてみてください。
実際に if 文とかをかいてみるといいかな。

a5ob7r replied:
まあ、結局はこうなると思います。でも、#travel_toが無ければ、こうやって外部から時間を指定できるようにしていたと思います。

```ruby
if token.expires_at? Time.current
  ...
end

# とか

def expired?
  expires_at? Time.current
edn
```

tk0miya replied:
書いてもらったように、 `#expires_at?` というメソッドを定義するということは外部から日時を渡して判定させたいということですよね。現在時刻だけでなく、未来の時刻や過去の時刻をわたして呼び出したいということだとおもうのですが、あっていますか?

僕はどういう場面で未来の時刻や過去の時刻を渡すのか、というのがあまりピンときていないのですが、なにか想定している用途があるのでしょうか?

epaew replied:
選択肢としては
```ruby
def expired?(at = Time.current)
def expired?(at: Time.current)
```
ってのもありますよ？

a5ob7r replied:
> 書いてもらったように、 `#expires_at?` というメソッドを定義するということは外部から日時を渡して判定させたいということですよね。現在時刻だけでなく、未来の時刻や過去の時刻をわたして呼び出したいということだとおもうのですが、あっていますか?
> 
> 僕はどういう場面で未来の時刻や過去の時刻を渡すのか、というのがあまりピンときていないのですが、なにか想定している用途があるのでしょうか?

もっぱらテストでの使用を想定していました。#travel_toが無ければ、Time.current等の副作用ありのメソッドを利用すると、テストしづらいのではと考えていました。

tk0miya replied:
>もっぱらテストでの使用を想定していました。

この一言がないのですれ違ってましたね。
僕は正式な (アプリに公開する) メソッドとして `#expires_at?` を作ろうとしているのかと思っていました。
上の方の自分のコメントを読み返してみてください。自分の考えを相手に伝えきれているでしょうか。
議論は相手と意見が噛み合ってから、ようやくスタートするものです。

自分の考えを伝えるというのは簡単なことではないのですが、こういう些細なかけちがいで消耗するのはもったいないので、今後も頑張っていきましょう。


さて、テストで使うという想定だという認識が揃ったところで、次の話をしていきましょうか。
僕が考えるポイントはこのあたりです。

* アプリ内で使うメソッドとして `#expires_at?` が適当か
* アプリでは `#expired?` を使って、裏口として `#expires_at?` を提供するのか
* epaew がコメントしてくれた optional 引数を使う手はどうか
* Rails では (ActiveSupport では) `#travel_to` を提供しているが、裏口は残しておいたほうが良いか?

他の言語では `#travel_to` みたいな仕組みは (公式では) 提供されていないので、引数で時刻を渡すという口を設けるというアプローチはよくやるのですが、Ruby on Rails では公式で提供しているのでそれを使わない選択をするのか、というのが一番の焦点になるのでは、と思います。

a5ob7r replied:
承知しました。

epaew commented on app/models/access_token.rb:
[IMO] （別 PR で良さそう）

「有効なトークンの中で一致するものはあるか？」が多分今後欲しくなる気がします。
となると `scope :alives, -> { ... }` があると便利じゃないかな？

a5ob7r replied:
ありがとうございます。考えておきます。

a5ob7r committed:
Fix the term to indicate revocations

The term was the same to the term to indicate expirations. This is too
bad situation and causes confusion for me.

a5ob7r committed:
Remove unnecessary #travel_to

Use #change matcher to test value modifications by some actions.

a5ob7r committed:
Move #travel_to to a before block

This change is for the consistency with the test case explanation.

a5ob7r committed:
Remove an unnecessary validation

a5ob7r committed:
Change #expires? to #expired

The concern of the method is that the token is already expired.

a5ob7r committed:
Add a method to User to make a new access token

a5ob7r commented:
一通り修正が完了しました。レビューをしていただけますと幸いです。

tk0miya commented on app/models/user.rb:
```diff
@@ -1,11 +1,19 @@
   has_secure_password validations: false
   validates :password, presence: true, length: { minimum: 16, maximum: 128 }
+
+  def new_access_token(token: SecureRandom.base64, scope: 'READ WRITE', expires_in: 1.hour)
```
AccessToken の `#initialize` メソッドに書く方が便利そうです。

そうすると
```
AccessToken.new(...)
```
の場合も
```
user.access_tokens.build(...)
```
の場合もデフォルト値が設定されますよね

後者は Rails の association の基本形なので、標準と同じやり方でトークンが生成できるのはいいですよね。

それに、トークンの生成ルールを知ってるのが User クラスってなんか責任分担がズレててもったいなくないですか?

a5ob7r replied:
> user.access_tokens.build(...)

おー、こういう書き方ができるんですね。それだと、確かに#initializeメソッドを上書きするのが良さげですね。#initializeの上書き自体は考えたのですが、デフォルトのメソッドを上書きするほどの理由が見つけられなかったので、採用しませんでした。でも↑なら、十分理由になりそうです。

> それに、トークンの生成ルールを知ってるのが User クラスってなんか責任分担がズレててもったいなくないですか?

これについてはどうなんですかね。考えなかったわけではないのですが、ユーザはトークンと関連を持っているので、それに関するルールを保持していてもそこまでおかしくないのでは、と考えていました。

あとはユーザにアクセストークンを保持しているというイメージから、ユーザに対するメソッド呼び出しでトークン作成をするようにしたかったということもありました。ただ、これについても↑の関連付けへのメソッド呼び出しで実現できるということなので、やはりそちらを採用したいと考えています。

tk0miya replied:
>おー、こういう書き方ができるんですね。それだと、確かに#initializeメソッドを上書きするのが良さげですね。

このあたりに has_many association によって生えてくるメソッドの一覧がありますよ

https://railsguides.jp/association_basics.html#has-many%E9%96%A2%E9%80%A3%E4%BB%98%E3%81%91%E3%81%AE%E8%A9%B3%E7%B4%B0%E3%81%AA%E5%8F%82%E7%85%A7

>これについてはどうなんですかね。考えなかったわけではないのですが、ユーザはトークンと関連を持っているので、それに関するルールを保持していてもそこまでおかしくないのでは、と考えていました。

この考え方を進めていくと、複雑なアプリケーションではユーザーはいろんなオブジェクトの初期化ルールを持つことになりますね。

* 音楽のプレイリスト
* ECサイトのカート
* 購入履歴
* 視聴履歴
* などなど

Amazon や Google といった巨大サービスの User クラスの持つべき初期化メソッドを考えてみると良さそうです。

あとでコードを追いかける際も、AccessToken の作り方を調べる際には最初に AccessToken クラスを見ることがおおいと思います。そして、記述がないので実際に作っている箇所 (controller など) を見に行って… という順序になりそうです。
あとでコードを置いやすいという点でも、特別な理由がなければ各クラス側で初期化ルールをもっていたほうがシンプルになると思います。

a5ob7r replied:
初期値の設定方法をいくつか調べてみたのですが、他にも何かあったりしますかね。

- DBレベルでの初期パラメータの指定(DBスキーママイグレーション)
    - 可能ならこれが一番確実である。
    - 変更が少し面倒。
    - 0だとか空文字だとか、静的な値しか設定できない。
- #after_initializeコールバックの使用
    - #initializeをオーバーライドしなくてもよい。
    - #new_record?などによる条件付けがないと、うっかり属性を上書きしてしまう可能性がある。
    - #initialize後に呼ばれるため、値を見てデフォルト値を設定してよいか判断する必要がある。
    - #initializeで引数で明示的にnilを設定していても、#after_initialize内ではそれが明示的に設定されたものなのかそうでないのかの判別がつかない。
- #before_validateコールバックの使用
    - 概ね#after_initializeコールバックと同様の利点と欠点がある。
    - #after_initializeより呼ばれる頻度が多い。
- Attributes APIの使用
    - オブジェクトレベルでモデルに属性を生やしたり、既存の属性を上書きするAPIをデフォルト値の設定だけに利用するのはやりすぎなような。
- #initializeメソッドのオーバーライド
    - 場合によってはオブジェクト生成時に呼ばれないこともある。
    - https://blog.dalethatcher.com/2008/03/rails-dont-override-initialize-on.html
    - #after_initializeでの問題であった、明示的なnil等の値の設定を検知可能である。
- defaults gemの利用
    - 内部的には#after_initializeと#new_record?による条件付けを利用している。
    - 宣言的にデフォルト値を設定可能である。
    - https://github.com/fnando/defaults
- 自身のクラスメソッドやメソッドとしてデフォルト値を利用した生成用メソッドの作成
    - 他の方法だと、デフォルト値を設定してほしくない場合の制御が難しい。コールバックの抑制とかは注意して利用する必要がある。でも、単なるメソッドとして実装するなら、それらを呼び分けるだけで済む。
    - 実装する適切な場所を考える必要がある。大抵は自分自身のクラスに寄せて実装するとよさげ。

tk0miya replied:
主なものはこれくらいですかねえ。

a5ob7r replied:
AccessTokenモデルの#initializeを上書きしてデフォルト値をよしなに設定するようにしました。この部分に関してはこれで問題ないでしょうか。

ちなみに、問題ないのであれば、テストの構成もデフォルト値を前提にしたものに変更しようと考えています。

a5ob7r replied:
明示的に引数を渡した場合は、例えnilが設定されたとしても属性を上書きしないようにしています。アプリケーション上でこれが発生することで困るような状況があるかについてはわからなかったです。ですが、nilを設定した場合のテストをする際、これが発生すると、一度#newを実行してオブジェクトを生成してから、再度該当する属性にnilを代入する羽目になるため、不便でした。なので、このような仕組みにしました。

あとは、個人の感覚の問題ですが、明示的に設定しているのにそれを暗示的に変更されるというのは違和感があります。

a5ob7r replied:
余談ですけど、デフォルト値でオブジェクトを作る場合のインターフェースでいうと、 `Class.new.with_defaults` みたいなのがいいなとは思いました。FactoryBotみたいに `Class.new(with_defaults)` みたいのもアリですね。とにかく、可能なら明示的にやりたいなとは思っていました。今回はできませんでしたが。

tk0miya replied:
>明示的に引数を渡した場合は、例えnilが設定されたとしても属性を上書きしないようにしています。アプリケーション上でこれが発生することで困るような状況があるかについてはわからなかったです。

ここが設計するポイントでは?
アプリケーション上で困ることがあるのであれば

* 後段のバリデーションではじく
* そもそも指定を許可しない

などの手があると思います。

先ほどコメントをしましたが、token を nil に設定したいことがあるのか、クラスの利用者に明示的に設定させたいのか、ですね。
利用者がトークン文字列を生成するということがないのであれば、指定不可にしてしまう手もありますよね。

僕らが作っているのは自由度の高いプラットフォームではなく、具体的な目的や仕様のあるアプリケーションです。ですから、開発者が任意のパラメータを指定してもうまく動くというのは求められていません。

※ むしろ、自由度が高すぎるとミスやバグが発生する原因にもなるので、制限をかけたほうが良い場面の方が多いです。

>ですが、nilを設定した場合のテストをする際、これが発生すると、一度#newを実行してオブジェクトを生成してから、再度該当する属性にnilを代入する羽目になるため、不便でした。なので、このような仕組みにしました。
>
>あとは、個人の感覚の問題ですが、明示的に設定しているのにそれを暗示的に変更されるというのは違和感があります。

たとえばなにかのカラムに nil を指定したらどうなるのか、どうなって欲しいのかというのを「クラス提供者」「クラス利用者」の両方の観点で考えてみてほしいです。
token でも scope でも expires_in でもなんでもよいのですが、どう動いてほしいのでしょうか。

tk0miya commented on spec/models/access_token_spec.rb:
```diff
@@ -0,0 +1,186 @@
+    it { is_expected.to be_valid }
+  end
+
+  describe '.token' do
```
以前は気にしていなかったけど、このあたりの describe ってバリデーションの動作確認ですよね。
なので、こういう describe を作って、その下に token, scope, ... とぶら下げるのが良さそう。
```suggestion
  describe '#valid?' do
```

※ テストは invalid? で行っているけど `#valid?` と `#invalid?` は対のものだし、伝わると思うので大目に見てもらおうw

tk0miya replied:
改めて今回 spec を読んでて、 `.token` という名前にもやっとしたんですよね。
Ruby で `.token` という書き方をすると、クラスメソッド名のことを指すので ??? となってしまったのです。

a5ob7r replied:
> Ruby で .token という書き方をすると、クラスメソッド名のことを指すので ??? となってしまったのです。

この辺りの慣習ってどこを見ればわかりますでしょうか。

a5ob7r replied:
実はちょうど今日、その辺りの話題を調べていました。なので、 `.method` がクラスメソッドを意味するということは知っていたのですが、これまでも `.attribute` の書き方をしていたけど突っ込まれなかったし、修正するにしてもこのPRではないよな、ということで保留にしていました。

https://www.betterspecs.org

属性についてテストする場合はどうやって書くものなのでしょうか。もしかして、そもそも属性に注目してテストは書かなかったりしますか。

tk0miya replied:
残念ながらどこで見たのかは覚えてないです。適当にググってみたところだとこのあたりでしょうか。
https://docs.ruby-lang.org/ja/latest/doc/symref.html#period

>これまでも .attribute の書き方をしていたけど突っ込まれなかったし、修正するにしてもこのPRではないよな、ということで保留にしていました。

はい、この PR でなくても良いと思います。
最初のコメントに書いたのですが、以前はあまり違和感を感じなかったので見落としていました。

そういう意味では「突っ込まれなかった」==「問題がひとつもない」ではないので、マージした後にやっぱりこっちのほうが良かった、ということはありえます。
レビューする側も人間なので、すべてのことについて完璧に指摘することは難しいので、そこは容赦してくださいw

>属性についてテストする場合はどうやって書くものなのでしょうか。もしかして、そもそも属性に注目してテストは書かなかったりしますか。

用語を揃えておいたほうが良さそうなので、最初に整理させてください。

* おそらく「属性」と呼んでいるものは「インスタンス変数」と「アクセサメソッド」(仮称)に分かれそうです。
* インスタンス変数というのは @ で始まる変数です (例: @foo, @bar)
* アクセサメソッドというのは attr_reader, attr_writer, attr_accessor で定義するメソッドのことで、インスタンス変数の読み書きをするメソッドを表します
    * 簡単にググった範囲ではこれらの総称を見つけられたので仮称としています

これらを踏まえた上で

* インスタンス変数そのもののテストは行いません。
    * オブジェクトの内部状態で基本的には触ることはできませんし、クラスの利用者は意識しなくてもよいものです。
* アクセサメソッドについてもテストはあまり行いません。
    * validates のところでも議論がありましたが、attr_reader で読み込みメソッドが生えていること、はテストをするメリットが薄いという考えです。

という回答になるかと思います。

一方で、今回コメントしている対象の部分は「インスタンス変数」や「アクセセサメソッド」のテストではないですよね? 提案したように、これらはバリデーションの(バリデーションルールのテストです。
ですから、バリデーションの判定関数である `#valid?` のテストとして記述するのがよいという考えです。
(各属性ごとにバリデーションを行っているので `#valid?` 配下の `token` 属性に関するテストという表現がよさそうだと思っています。

epaew replied:
> > Ruby で .token という書き方をすると、クラスメソッド名のことを指すので ??? となってしまったのです。
> 
> この辺りの慣習ってどこを見ればわかりますでしょうか。

私も「どこかに『こんな文化があります』が書いてあった」は出せないのだけれど。  
私は「るりま（Ruby リファレンスマニュアル、 https://docs.ruby-lang.org/ja/latest/doc/index.html ）がその書き方をしているから、それに倣っている」だと思っています。

- Class.new
    - https://docs.ruby-lang.org/ja/latest/method/Class/s/new.html
      > singleton method Class.new
- Class#new
    - https://docs.ruby-lang.org/ja/latest/method/Class/i/new.html
      > instance method Class#new

a5ob7r replied:
> 一方で、今回コメントしている対象の部分は「インスタンス変数」や「アクセセサメソッド」のテストではないですよね? 提案したように、これらはバリデーションの(バリデーションルールのテストです。
> ですから、バリデーションの判定関数である #valid? のテストとして記述するのがよいという考えです。
> (各属性ごとにバリデーションを行っているので #valid? 配下の token 属性に関するテストという表現がよさそうだと思っています。

あー、確かにそうですね。属性ごとにバリデーションを考えていたので、つい属性という観点で見ていましたが、実態は #valid? のテストですね。

あとインスタンス変数とアクセサメソッドはがっつり混同していました。失礼しました。

tk0miya replied:
雑談: 今後、複雑なアプリケーションを作る場合、複数の属性にまたがったバリデーションが出てくることがありますね。
〜〜権限の場合は〜〜が必須、とかそういうやつ。

そういうことも踏まえると、属性単位で定義をするのは幸せになれないですね

a5ob7r replied:
なるほど。

tk0miya commented on spec/models/access_token_spec.rb:
```diff
@@ -0,0 +1,186 @@
+    end
+  end
+
+  describe '#expires?' do
```
名前変え忘れですね

a5ob7r replied:
おっと。修正します。

a5ob7r committed:
Fix a typo of a method name in test description

a5ob7r committed:
Use custom initializer to make AccessToken instances to set defaults

tk0miya commented on app/models/access_token.rb:
```diff
@@ -0,0 +1,39 @@
+  def initialize(attributes = nil)
+    super attributes
+
+    return unless new_record? && attributes.is_a?(Hash)
```
ここにくるときって必ず new_record じゃないっけ

a5ob7r replied:
メソッドの説明では、必ずしもtrueではないように読み取れますが、#initialize内ではそうでもないんですかね。

https://api.rubyonrails.org/classes/ActiveRecord/Persistence.html#method-i-new_record-3F

tk0miya replied:
>Returns true if this object hasn't been saved yet – that is, a record for the object doesn't exist in the database yet; otherwise, returns false.

`AccessToken.new` して、オブジェクトが生成している最中にデータベースに保存されていることってあるのかな。
どんな場合に new_record じゃなくなるという想定をしていました?

a5ob7r replied:
DBに保存されているレコードを読み出してインスタンス化することを考えていました。でも、親メソッドを見てみたら、がっつりまだ保存されていないときに呼ばれると書いてありますね。実際、コードを見てみてもそのように見受けられました。

> New objects can be instantiated as either empty (pass no construction parameter) or pre-set with attributes but not yet saved (pass a hash with key names matching the associated table column names).

https://api.rubyonrails.org/classes/ActiveRecord/Core.html#method-c-new

というわけで僕の知識不足でした。失礼しました。

tk0miya replied:
脱線: DB に保存されているレコードを読み出す場合、どうやってるんだろうねえ。僕もあまり詳しくないので、今度時間が取れたら調べてみます。
おそらく new した後に (`#initialize` が動いたあとに) 読み込んだ属性や `@new_record` フラグを上書きしているんだと思いますが…

a5ob7r replied:
知識不足のほかに勘違いもありました。

#after_initializeは、.findの場合でも.newの場合でも、両方で呼ばれるとありました。

> Lastly an after_find and after_initialize callback is triggered for each object that is found and instantiated by a finder, with after_initialize being triggered after new objects are instantiated as well.

https://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html

なので、これを利用してデフォルト値を設定する場合は、#new_record?による判定を入れたほうがいいという話があったのですが、そちらと混同してしまいました。

https://qiita.com/supercell1023koeda/items/46aad5929491acd20108

a5ob7r replied:
> 脱線: DB に保存されているレコードを読み出す場合、どうやってるんだろうねえ。僕もあまり詳しくないので、今度時間が取れたら調べてみます。
> おそらく new した後に (`#initialize` が動いたあとに) 読み込んだ属性や `@new_record` フラグを上書きしているんだと思いますが…

こちらにはnewではなくallocateを使うとありますね。そして、まさに前述の僕の勘違いを正してくれる情報ですね。#after_initializeを調べていたときに見たはずなのですが。目の前に答えがあったのに見逃してました...。

https://blog.dalethatcher.com/2008/03/rails-dont-override-initialize-on.html

tk0miya replied:
見っけたよ、と書こうとしたらちょっと遅かったw
https://github.com/rails/rails/blob/main/activerecord/lib/active_record/persistence.rb#L632

a5ob7r commented on spec/models/user_spec.rb:
```diff
@@ -146,4 +146,24 @@
+    end
+  end
+
+  describe '#access_tokens' do
+    describe '#build' do
+      context 'デフォルト値で実行した場合' do
+        it '作成されたアクセストークンは有効である' do
+          expect(user.access_tokens.build).to be_valid
+        end
+      end
+    end
+  end
```
ちょうどRubocopで指摘されているところですが、このテストってどう書くのがいいんですかね。テストしたいこととしては、UserモデルがAccessTokenへの1:nの関連付けのオブジェクトを持っていること、 `user.access_tokens.build` で有効なAccessTokenオブジェクトを作成できることです。今の書き方だと、見た目上はホワイトテストになっていると思うのですが、何かいい方法はありませんでしょうか。

tk0miya replied:
rubocop の指摘はこれかな

```
spec/models/user_spec.rb:162:7: C: RSpec/NestedGroups: Maximum example group nesting exceeded [4/3].
      context 'デフォルト値で実行した場合' do
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

※ CI から指摘は辿れるのだけど、貼ってもらったほうが遷移が少なくて読む側には助かりますねー

tk0miya replied:
おっと、途中で送信してしまった。

rubocop の指摘は特に関係なくて、テストの書き方が気になっているということであっています?

>テストしたいこととしては、UserモデルがAccessTokenへの1:nの関連付けのオブジェクトを持っていること、 user.access_tokens.build で有効なAccessTokenオブジェクトを作成できることです。

これって has_many が正しく動作しているか(宣言されているか)のテストをしたいってことでいいのかしら。
バリデーションの話と同じく、これはあまりテストしませんね。
いずれも ActiveRecord の機能なので、機能的な確認にはあまり意味がありません。

その上で、どう確認しようかという技術的興味を満たすのだとするとこんな感じでしょうか。

* UserモデルがAccessTokenへの1:nの関連付けのオブジェクトを持っている
    * User データと AccessToken データ(複数)をそれぞれ作って、 `user#access_tokens` でそれらが取得できることを確認するとか?
* user.access_tokens.build で有効なAccessTokenオブジェクトを作成できること
    * 今の実装でよさそう

a5ob7r replied:
考え直した結果、仕様として明示する必要は薄いと判断したため、該当箇所は削除しました。

tk0miya commented on app/models/access_token.rb:
```diff
@@ -0,0 +1,39 @@
+
+    return unless new_record? && attributes.is_a?(Hash)
+
+    set_default = ->(name, default) { self[name] ||= default unless attributes.key? name }
```
これ、単に
```
self.token ||= ...
```
でいいと思うよ。

a5ob7r replied:
それだと、明示的にnilを指定したときでも上書きされてしまいませんか。

tk0miya replied:
あー、ごめん。説明が足りなかった。nil を指定したいことがあるんだっけ? というのがそもそもの疑問にあった上で、上の書き方で良いのでは、という提案でした。
前提を端折っていたので良くないコメントでしたね、ごめん。

※ nil を指定したいことがあるんだっけ〜 の部分は別コメントで書いたとおりです。

a5ob7r replied:
試してみたところ、RailsのAttribute APIによるデフォルト値の設定を使うと、明示的にnilを設定した場合は、ちゃんとnilを設定してくれるようです。なので、この仕様を実現するためにわざわざ#initializeをオーバーライドする必要はないようです。

tk0miya commented on app/models/access_token.rb:
```diff
@@ -0,0 +1,39 @@
+  validates :expires_in, numericality: { greater_than_or_equal_to: 0 }
+
+  # 新規レコードの場合かつ属性が引数で明示的に設定されていない場合はデフォルト値を設定する。
+  def initialize(attributes = nil)
```
デフォルト引数を Hash にしたらいろいろはかどるのでは?

```suggestion
  def initialize(attributes = {})
```

a5ob7r replied:
親と思われる関数ではデフォルト値がnilだったので、そこに合わせました。

https://api.rubyonrails.org/classes/ActiveRecord/Core.html#method-c-new

epaew replied:
多分それなりに新しい Ruby 処理系使って今の開発って進めてもらってますよね？

↓の書き方ってどこかで deprecated になってなかったかしら。
```ruby
def initialize(attributes = {})
```
↓のほうが最近っぽい気が
```ruby
def initialize(**attributes)
```

epaew replied:
この辺りですかね
- https://docs.rubocop.org/rubocop/cops_style.html#styleoptionhash
- https://rubystyle.guide/#keyword-arguments-vs-option-hashes

tk0miya replied:
つい古い Ruby 処理系をサポートしたくなる癖が出たけど、いまなら `**attributes` 一択ですね…

※ ツールとかフレームワークを作る側なので、ひろく対応したくなっちゃうマンです

a5ob7r replied:
↑のような.initializeのオーバーライド時のデフォルト引数の変更を試みたのですが、現時点では変更は事実上不可能というオチでした。

というのも、ActiveRecord::Baseを継承しているクラスが.newを呼ぶと、以下のメソッドを呼ぶことになります。そして、このメソッドのデフォルト引数がnilに設定されており、かつその引数が最終的に子クラスの.initializeに渡されると思われます。本当は、ここについてはまだコードを追いきれていないですが、実験やデバッガで確認した範囲だと、こういう理解でよさそうでした。つまり、.newを引数省略で呼んだとしても、呼び出し順の上位のメソッドのデフォルト引数であるnilが渡されることになるので、↑のように.initializeメソッドのオーバーライド時にデフォルト引数を設定しても意味がないということでした。

https://github.com/rails/rails/blob/e446ef82c8174a4bd6c98f670912bb395b3091d6/activerecord/lib/active_record/inheritance.rb#L56

なので、もし.initializeメソッドをオーバーライドする必要があるのなら、デフォルト引数を設定してもほとんど意味はないが、引数省略時にnilが渡されるという挙動を暗示できるため、オーバーライド時でもnilを設定しておくといいかもしれない、ということでした。

あと、**演算子を利用したメソッド定義に関しては、上記のメソッドが旧来のスタイルを利用して定義されている都合上、キーワード引数ではなく一つのハッシュとして引数を渡してきます。よって、**演算子を利用してメソッド定義をしていると、新しいRubyの処理系ではArgumentErrorが発生します。というわけで、こちらの方法も.initializeのオーバーライド時には利用できないということのようです。

tk0miya replied:
調査 & 実験ありがと。なるほど、これはやってみないとわからないやつですね。僕も知らなかったので振り回してしまってごめん。
(たぶん僕なら initialize の中で上書きしてしまうやりかたを取るので…)

tk0miya commented on app/models/access_token.rb:
```diff
@@ -0,0 +1,39 @@
+
+    set_default = ->(name, default) { self[name] ||= default unless attributes.key? name }
+
+    set_default.call :token, SecureRandom.base64
```
外部からトークンを指定したいことってあります?
変なトークンを作らせないために、外部から指定できないようにするというのも手だと思います。

a5ob7r replied:
今回は、トークンのフォーマットをバリデーションで制限しているので、変なトークンを指定しても保存はされないようになっていると思います。あとは、外部から指定できるとテストのときは便利そうではありますね。実際、今のテストで指定しているところがあります。

tk0miya replied:
外部から指定できると嬉しいのは重複チェックのときぐらいしょうか。あってます?
その一箇所だけであれば、指定できないようにしてしまったほうがミスが減るのでは、という感じがしました。

バリデーションがあるから保存されない、というよりもそもそも一般の開発者にはトークン文字列を生成させたくないと僕は考えています。
このクラスの開発者はアクセストークンの中身について意識しなくて良いというという利便性の観点、
安全なアクセストークンを生成する方法を全員が知っているわけではないというセキュリティの観点あたりから、
外部からトークン文字列を指定させるのはお得じゃないなあという感じですねえ。

ちなみに、反対の考え方として

* 指定できる口があるからといって、みんなが使うわけではないので気にしない (使わないはずでしょ論)
* わざわざ防御的な実装をするコストがもったいない
* テストなど、例外ケースでたまに使うので必要である。基本的には使わせないルールにする

あたりがありそうです。

※ 絶対に入り口を塞ぐべきだと主張したいわけではなく、誰がどういうときに指定するのかと言う観点が必要そうだと思って話を広げています。nil を指定した場合の動きの話などがでてきて、利用のシチュエーションを僕が掴みかねているのです。

a5ob7r committed:
Fix the format to validate the scope attribute

KT-tim commented on spec/models/access_token_spec.rb:
```diff
@@ -0,0 +1,198 @@
+    end
+  end
+
+  describe '.scope' do
```
validなscopeのテストも必要だと思います

a5ob7r replied:
必要な範囲で正常系のテストを追加しました。

a5ob7r committed:
Add some missing not null constrains to AccessToken

a5ob7r committed:
Change specs for #new and #valid? of AccessToken

a5ob7r committed:
Add subjects for each of contexts to indicate the test target

a5ob7r committed:
Add a no argument test to indicate `user` is required

a5ob7r committed:
Add some tests about default values of AccessToken

a5ob7r committed:
Change descriptions for some tests

a5ob7r committed:
Remove some unnecessary validation tests

These are not critical concerns about this model or there are the same
tests in another context.

a5ob7r committed:
Sort a test

a5ob7r committed:
Change a test description

a5ob7r committed:
Add some positive tests for scope

a5ob7r committed:
Remove the global subject

a5ob7r committed:
Add a test about #expired? call if not saved yet

a5ob7r committed:
Relax Max of RSpec/NestedGroups for Model Specs

Too many nested groups may lead to unreadable specs, so we should try to
reduce them as possible. However if these are well structured, this
constraint is a little bit tight. So we relax it only for Model Specs
because they are well structured as below.

  Model -> Method -> Context* -> It

a5ob7r committed:
Add a missing subject

a5ob7r commented:
AccessTokenモデルのスペックに関して修正したので、該当箇所(`spec/models/access_token_spec.rb`)のレビューをお願いします。

今回はスペックがテストの羅列ではなくドキュメントであると意識して記述しました。

- メソッド単位でのテストのグループ化(.new, #valid?)
- アプリケーションのロジックに関してあまり重要ではないテストの削除
- モデルに関するスペック(仕様書、ドキュメント)として必要な情報をテストとして追加

tk0miya commented on app/models/access_token.rb:
```diff
@@ -0,0 +1,41 @@
+  # 上記では、トークンの文字列長は1以上であればよいことになっているが、短すぎると総当たりで突破されないとも
+  # 限らないので 保険として、 `=` によるパディング以外の部分に関して、ある程度の長さを要求する。
+  validates :token, format: { with: %r{\A[\w\-.~+/]{16,}=*\Z} }, uniqueness: true
+  validates :scope, format: { with: /\A(READ|WRITE)( (READ|WRITE))*\Z/ }
```
このバリデーションルールだと `READ READ` とか `WRITE WRITE` も OK になっちゃいますね。

`format` 指定でできることを超えてそうなので、カスタムバリデーションを定義するのが良さそう。

※ この PR の議論は長引いているので、この話は対応するとしても別 PR にするのがよさそう

a5ob7r replied:
そのこと自体は把握していました。ただ、現時点では困る状況が特に思い当たらなかったので、修正は考えていませんでした。

tk0miya replied:
認識の上で雑な実装にしているのであれば OK です (トレーニングのメインではないですし)

tk0miya commented on spec/models/access_token_spec.rb:
```diff
@@ -0,0 +1,181 @@
+        it { is_expected.to be_valid }
+      end
+
+      context '適当な値のとき' do
```
適当な、というのは意味が広いので使わないほうがいいですね。辞書を引くとふさわしい、当てはまっているというのを指すので、口語で使う意味と正反対になります。
https://www.weblio.jp/content/%E9%81%A9%E5%BD%93

a5ob7r replied:
承知しました。修正します。

tk0miya commented on spec/models/access_token_spec.rb:
```diff
@@ -0,0 +1,181 @@
+    describe 'expires_in' do
+      subject { described_class.new expires_in:, user: }
+
+      context '0のとき' do
```
テストとしてはこの記述で正しい (境界値をちゃんとテストしている) のですが、仕様としてはちょっと理解しづらいですね。

* 0以上の値の場合
* マイナス値の場合

という context にするのがよいと思います。

もう少しテストっぽさを加えるのであれば

* 正数の場合
* 0 の場合
* 負数の場合

という 3ケースに分けるのもよさそうです。

a5ob7r replied:
それもそうですね。今回は前者を採用したいと思います。

tk0miya commented on app/models/access_token.rb:
```diff
@@ -0,0 +1,41 @@
+  end
+
+  def expired?
+    raise 'Save this object before call #expired?' if new_record?
```
質問: このチェックってなんで必要なんでしたっけ? 特に塞ぐ必要もなさそうな気がしました。

a5ob7r replied:
#expired?内部でcreated_at属性を参照するので、この値がnilになる場合、例えばオブジェクトが永続化されていない場合だと、確かにこのガード文がなくてもNoMethodErrorが発生するので、例外が発生するという意味では、あってもなくても結果は同じです。

ただ、そのような状況での呼び出しを想定していないということを明示できるので、先のガード文をあえて記述しました。

tk0miya replied:
ああ、なるほど、そういうことか。納得しました。

tk0miya commented on spec/models/access_token_spec.rb:
```diff
@@ -0,0 +1,181 @@
+  describe '#revoked?' do
+    subject { described_class.new revoked_at:, user: }
+
+    context '無効化されていたら' do
```
雑談: context は〜〜の場合、〜〜のとき、と書くと収まりがいいですよ。
文語体で書いたほうがちょっとかっこいいですw

a5ob7r replied:
他の箇所は文語で書いていたのですが、この周辺だけ違いますね。修正しました。

tk0miya commented on spec/models/access_token_spec.rb:
```diff
@@ -0,0 +1,181 @@
+    end
+
+    context '無効化されていなかったら' do
+      before do
+        subject.revoke
+      end
```
この before 節あってますか? 無効化しているように見えるのですが…

a5ob7r replied:
ここはややこしいかもしれませんね。ここでは#revokeの実行による副作用がテスト対象です。つまりは、事前に#revokeを実行する必要があるわけで、それをitのブロック内部で#revokeではなくbefore文のブロックで実行しています。

ただ、ここに関しては、書いた本人ですら毎度「あれ？」と思うくらいには誤解しやすい箇所だと思っています。修正できそうだったらやってみます。

a5ob7r replied:
changeマッチャを使うように変更しました。副作用のテストにはこちらを使うべきでしたね。

tk0miya replied:
changes マッチャを使うことで違和感の正体が浮かび上がってきました。
ここ全体のグループのテスト対象は `AccessToken#revoke` なのですから、以下のような subject になるのが正しいのでは? subject == テスト対象を指すのが適当です。
```
subject { access_token.revoke }

let(:access_token) { AccessToken.new ... }
```

subject がズレているので、他の記述がズレていったということだと思います。

「revoke されたトークン」に対して `#revoke` を呼ぶという場合であれば「revoke されたトークン」を作るのは let や before で行うのが適当ですが、 `#revoke` を呼ぶというのは subject を使うのが適当ですね。

a5ob7r replied:
なるほど、修正します。

tk0miya reviewed (commented):
とりいそぎ spec のところだけ見ました

a5ob7r committed:
Fix how to test the side effect for AccessToken#revoke

a5ob7r committed:
Add a missing space

a5ob7r committed:
Fix some context descriptions with literal languages

a5ob7r committed:
Change some subjects to target method calls

a5ob7r committed:
Fix some test description about expires_in

a5ob7r committed:
Fix an ambiguous term in the test description

a5ob7r committed:
Remove a test to change the spec to be undocumented

This value is valid now, but there is no good reason to support
duplicate values. This should be removed in the future.

a5ob7r committed:
Add a test about space-delimited list order for scope

a5ob7r committed:
Remove an unnecessary conditional

Because new_record? is always true in .initialize.

a5ob7r committed:
Remove an unnecessary test for User

This test is not very important for User because this feature is from
ActiveRecord. This is a white-box test.

a5ob7r committed:
Switch to defaults gem to set default values

We can distinguish a nil value is set implicitly or explicitly for the
attributes. This behavior is ideal as a program, but probably nobody
will be in trouble even if no this behavior because these values
disallow nil value and we end up setting not nil to these values.

a5ob7r committed:
Switch to Rails Attribute API to set default values

This is a Rails native way and provides a very simple feature for
default values. It's enough for us.

a5ob7r committed:
Remove unnecessary attributes for FactoryBot of AccessToken

This model has default values for these attributes.

a5ob7r commented:
一通り修正が完了しました。レビューをお願いいたします。

tk0miya reviewed (approved):
+1 です。おつかれさまでしたー

a5ob7r commented:
ありがとうございました。
